ca_tools/scripts/controller.py

```diff
@@ -0,0 +1,80 @@
+#! /usr/bin/env python
+# -*- coding: utf-8 -*-
+
+
+import math
+from RobotLocalizationTf import RobotLocalizationTf
+
+import rospy
+from geometry_msgs.msg import Twist, Pose, Quaternion, PoseWithCovariance
+from nav_msgs.msg import Odometry
+from std_msgs.msg import Bool
+from tf.transformations import euler_from_quaternion as efq
+
+
+
+
+class PoseController():
+
+    def __init__(self, velocity_publisher_topic = 'cmd_vel'):
+        self._current_pose = Pose()
+        self._goal_pose = Pose()
+        self._pub_cmd = rospy.Publisher(velocity_publisher_topic, Twist, queue_size=10)
+        self._hz = rospy.get_param('~hz', 10)
+        self._rate = rospy.Rate(self._hz)
+        self._tf = RobotLocalizationTf()
+        self._kp = 0.1
+        self._ki = 0.0
+        self._kd = 0.0
+        self._turned_on = False
+        self._angle_diff = 0
+        self._rate.sleep()
+
+    def _current_pose_callback(self,data):
+        self._curent_pose = data
+        _, _, self._angle_diff,_ = self._tf.get_pose_diff(self._current_pose,self._goal_pose)
+
+    def _goal_pose_callback(self,data):
+        self._goal_pose = data
```

Comment:
You should use [locks](https://answers.ros.org/question/28373/race-conditions-in-callbacks/?answer=28386#post-id-28386) to protect `data` and information being called in parallel.
I mean, `self._goal_pose` is being set here and used somewhere else. Beware of that. The same applies for other callbacks.